sensors/mag/lsm9dsxx.c

```diff
@@ -0,0 +1,249 @@
+#define UTSL_2_MAGSHUB    10.F         /* microtesla to sensorhub magnetic flux storage scale of 10^-7 T */
+
+
+typedef struct {
+	spimsg_ctx_t spiCtx;
+	sensor_event_t evt;
+	char stack[512] __attribute__((aligned(8)));
+} lsm9dsxxMag_ctx_t;
+
+
+static int16_t translateMag(uint8_t hbyte, uint8_t lbyte)
+{
+	int16_t val = 0;
+	val |= hbyte;
+	val <<= 8;
+	val |= lbyte;
+	return val * MAG4G_CONV_MGAUSS * MGAUSS_2_UTSL * UTSL_2_MAGSHUB; /* value -> milligauss -> microtesla -> sensorhub storage scale*/
+}
+
+
+static int spiWriteReg(spimsg_ctx_t *spiCtx, uint8_t regAddr, uint8_t regVal)
+{
+	unsigned char cmd[2] = { (regAddr & 0x7F), regVal }; /* write bit set to regAddr */
+
+	return spimsg_xfer(spiCtx, cmd, sizeof(cmd), NULL, 0, sizeof(cmd));
+}
+
+
+static int lsm9dsxxMag_whoamiCheck(spimsg_ctx_t *spiCtx)
+{
+	uint8_t cmd, val, err;
+
+	cmd = REG_WHOAMI | SPI_READ_BIT;
+	val = 0;
+	err = spimsg_xfer(spiCtx, &cmd, 1, &val, 1, 1);
+	if ((err < 0) | (val != REG_VAL_WHOAMI)) {
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int lsm9dsxxMag_hwSetup(spimsg_ctx_t *spiCtx)
+{
+	if (lsm9dsxxMag_whoamiCheck(spiCtx) != 0) {
+		printf("lsm9dsxxMag: cannot read/wrong WHOAMI returned!\n");
+		return -1;
+	}
+
+	/* output data rate to 80Hz + high performance mode to X and Y axis */
+	if (spiWriteReg(spiCtx, REG_CTRL_REG1_M, (VAL_CTRL_REG1_M_XYHIGH_PERF_MODE | VAL_CTRL_REG1_M_ODR_80)) < 0) {
+		return -1;
+	}
+	/* scale +-4 gauss (same as default, written just for clarity) */
+	if (spiWriteReg(spiCtx, REG_CTRL_REG2_M, VAL_CTRL_REG2_M_FS_4) < 0) {
+		return -1;
+	}
+	/* high performance mode for Z axis */
+	if (spiWriteReg(spiCtx, REG_CTRL_REG4_M, VAL_CTRL_REG4_M_ZHIGH_PERF_MODE) < 0) {
+		return -1;
+	}
+	usleep(10 * 1000);
+
+	/* continuous measurement mode */
+	if (spiWriteReg(spiCtx, REG_CTRL_REG3_M, (VAL_CTRL_REG3_M_MD_CONTINUOUS | VAL_CTRL_REG3_M_SIM_SPIRW | VAL_CTRL_REG3_M_I2CDSBL)) < 0) {
+		return -1;
+	}
+	usleep(100 * 1000);
+
+	return 0;
+}
+
+
+static void lsm9dsxx_threadPublish(void *data)
+{
+	int err;
+	time_t tstamp;
+	sensor_info_t *info = (sensor_info_t *)data;
+	lsm9dsxxMag_ctx_t *ctx = info->ctx;
+	spimsg_ctx_t *spiCtx = &ctx->spiCtx;
+	uint8_t ibuf[SENSOR_OUTPUT_SIZE] = { 0 };
+	uint8_t obuf;
+
+	while (1) {
+		usleep(12500); /* 80Hz sampling period */
+
+		obuf = REG_DATA_OUT | SPI_READ_BIT | SPI_AUTOADDRINCR_BIT;
+		err = spimsg_xfer(spiCtx, &obuf, 1, ibuf, SENSOR_OUTPUT_SIZE, 1);
+		gettime(&tstamp, NULL);
+
+		if (err >= 0) {
+			/* minus accounts for non right-handness of lsm9dsxx accelerometer frame of reference */
+			ctx->evt.mag.magX = translateMag(ibuf[1], ibuf[0]);
+			ctx->evt.mag.magY = translateMag(ibuf[3], ibuf[2]);
+			ctx->evt.mag.magZ = translateMag(ibuf[5], ibuf[4]);
+			ctx->evt.timestamp = tstamp;
+			/* TEST */
+			printf("%d %d %d\n", ctx->evt.mag.magX, ctx->evt.mag.magY, ctx->evt.mag.magZ);
+			sensors_publish(info->id, &ctx->evt);
+		}
```

Comment:
What about calling `gettime()` only, when there is no error?
```suggestion
		if (err >= 0) {
			gettime(&tstamp, NULL);
			ctx->evt.timestamp = tstamp;

			/* minus accounts for non right-handness of lsm9dsxx accelerometer frame of reference */
			ctx->evt.mag.magX = translateMag(ibuf[1], ibuf[0]);
			ctx->evt.mag.magY = translateMag(ibuf[3], ibuf[2]);
			ctx->evt.mag.magZ = translateMag(ibuf[5], ibuf[4]);
			/* TEST */
			printf("%d %d %d\n", ctx->evt.mag.magX, ctx->evt.mag.magY, ctx->evt.mag.magZ);
			sensors_publish(info->id, &ctx->evt);
		}
```

Reply:
called `gettime` directly on event timestamp